### ISSUE TITLE
Inconsistent behaviors when using multiple trainer threads (with same actual minibatch size).

### ISSUE THREAD

**Issue by xinghai-sun** [PaddlePaddle/Paddle#2186]
`batch_size = 5, trainer = 8` and ` batch_size = 40, trainer = 1` should be expected to have similar (or the same, except for the sample order) computation results, since their actual minibatch size are both 40.

However,  the latter one runs into "Floating Point Exception" error, while the former one runs just ok, unexpectedly.

Is it a potential bug? Or something else might also be different between these two cases, causing such an inconsistency ?

**Comment by hedaoyuan** [comment]
Should be similar, but not necessarily the same.
> However, the latter one runs into "Floating Point Exception" error
@xinghai-sun Can you give me some data so that I can reproduce the problem, or paste the call stack when floating point exception occurs.

**Comment by xinghai-sun** [comment]
@hedaoyuan  Thanks for the reply. 

It occurs when running the demo in [machine translation chapter](https://github.com/PaddlePaddle/book/blob/develop/08.machine_translation/train.py) in PaddlePaddle Book. The same thing happens when running this [NMT with NTM](https://github.com/xinghai-sun/models/blob/3ff749b4b49a28c8eee66ba7eca03a4476605eb2/mt_with_external_memory/mt_with_external_memory.py) model.

I know that the error might be due to the potential bugs in gradient-clipping (I guess that Adam's second moment computation makes Floating Point Exception even more frequent to occur without proper gradient clipping).  

What I‘m curious about is not this bug itself (someone else might be working on it?), but the phenomenon that when I increased the trainer number, the error just disappeared.

What about difference in ADAM computation or clipping operations between these two cases?

```
Pass 0, Batch 0, Cost 244.581543, {'classification_error_evaluator': 1.0}
.........
Pass 0, Batch 10, Cost 252.381152, {'classification_error_evaluator': 0.9234693646430969}
.........
Pass 0, Batch 20, Cost 270.056104, {'classification_error_evaluator': 0.913250744342804}
.........
Pass 0, Batch 30, Cost 282.266187, {'classification_error_evaluator': 0.9371011853218079}
.*** Aborted at 1495084196 (unix time) try "date -d @1495084196" if you are using GNU date ***
PC: @                0x0 (unknown)
*** SIGFPE (@0x7fb8ddce2716) received by PID 14543 (TID 0x54b3d960) from PID 18446744073135859478; stack trace: ***
    @       0x3f0b90c420 (unknown)
    @     0x7fb8ddce2716 paddle::AssignCpuEvaluate<>()
    @     0x7fb8ddce3770 paddle::AssignEvaluate<>()
    @     0x7fb8ddcde240 paddle::adamApply()
    @     0x7fb8ddcdb296 paddle::AdamParameterOptimizer::update()
    @     0x7fb8ddcc7d55 paddle::SgdThreadUpdater::threadUpdateDense()
    @     0x7fb8ddcc8d6f _ZNSt17_Function_handlerIFvimEZN6paddle16SgdThreadUpdater11finishBatchEfEUlimE_E9_M_invokeERKSt9_Any_dataim
    @     0x7fb8dc76938c _ZNSt6thread5_ImplISt12_Bind_simpleIFZN6paddle14SyncThreadPool5startEvEUliE_mEEE6_M_runEv
    @     0x7fb8dbbad2f0 execute_native_thread_routine
    @       0x3f0b90610a (unknown)
    @       0x3f0b0c5ee3 (unknown)
Floating point exception (core dumped)
```
